SVGLengthValue.cpp should support rem (probably other) units for width and height attributes

nullhook opened the pull request:
#### 5e29e18f12fb9dc7245a7f03867364eb9fa5f9d6
<pre>
SVGLengthValue.cpp should support rem (probably other) units for width and height attributes
<a href="https://bugs.webkit.org/show_bug.cgi?id=258003">https://bugs.webkit.org/show_bug.cgi?id=258003</a>

Reviewed by NOBODY (OOPS!).

* Source/WebCore/svg/SVGLength.h:
* Source/WebCore/svg/SVGLengthContext.cpp:
(WebCore::SVGLengthContext::convertValueToUserUnits const):
(WebCore::SVGLengthContext::convertValueFromUserUnits const):
(WebCore::SVGLengthContext::convertValueFromUserUnitsToREMS const):
(WebCore::SVGLengthContext::convertValueFromREMSToUserUnits const):
* Source/WebCore/svg/SVGLengthContext.h:
* Source/WebCore/svg/SVGLengthValue.cpp:
(WebCore::lengthTypeToString):
(WebCore::parseLengthType):
(WebCore::primitiveTypeToLengthType):
(WebCore::lengthTypeToPrimitiveType):
* Source/WebCore/svg/SVGLengthValue.h:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5c4dc51c807c5473c5d92ea4b93a35b54e8b8da8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96846 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16460 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/6643 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101919 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47366 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/98891 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16756 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24905 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73798 "Failure limit exceed. At least found 470 new test failures: accessibility/aria-hidden-subtree.html accessibility/aria-hidden-with-elements.html accessibility/aria-hidden.html accessibility/aria-roles.html accessibility/css-content-attribute.html accessibility/heading-title-includes-links.html accessibility/image-link.html animations/animation-controller-drt-api.html compositing/absolute-inside-out-of-view-fixed.html compositing/backing/backing-store-attachment-animated-transform-inside-fixed.html ... (failure)") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31007 "Found 60 new test failures: animations/animation-controller-drt-api.html css3/flexbox/csswg/flexbox_direction-row-reverse.html css3/masking/clip-path-on-inline.html css3/unicode-bidi-float-crash.html editing/caret/caret-position-sideways-lr.html editing/caret/caret-position-vertical-rl.html editing/execCommand/5700414-1.html editing/execCommand/5700414-2.html editing/execCommand/format-block-with-trailing-br.html editing/execCommand/move-up-down-should-skip-hidden-elements.html ... (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99849 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12619 "Found 60 new test failures: accessibility/display-contents/search-traversal.html accessibility/ios-simulator/focus-change-notifications.html accessibility/ios-simulator/heading-text-updates.html accessibility/ios-simulator/ios-search-predicate.html accessibility/ios-simulator/textfield-in-axvalue.html accessibility/search-traversal-after-role-change.html animations/animation-controller-drt-api.html compositing/accelerated-layers-after-back.html compositing/backing-store-attachment-1.html compositing/backing/backing-store-attachment-animated-transform-inside-fixed.html ... (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/87610 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54132 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12379 "Found 60 new test failures: imported/w3c/web-platform-tests/FileAPI/idlharness.html imported/w3c/web-platform-tests/accname/basic.html imported/w3c/web-platform-tests/accname/name/comp_host_language_label.html imported/w3c/web-platform-tests/accname/name/comp_label.html imported/w3c/web-platform-tests/accname/name/comp_name_from_content.html imported/w3c/web-platform-tests/accname/name/comp_name_from_content.tentative.html imported/w3c/web-platform-tests/accname/name/comp_text_node.html imported/w3c/web-platform-tests/content-security-policy/base-uri/base-uri_iframe_sandbox.sub.html imported/w3c/web-platform-tests/content-security-policy/generic/generic-0_1-img-src.html imported/w3c/web-platform-tests/content-security-policy/generic/generic-0_1-script-src.html ... (failure)") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5430 "Found 60 new test failures: accessibility/aria-hidden-subtree.html accessibility/aria-hidden-updates-alldescendants.html accessibility/aria-hidden-with-elements.html accessibility/aria-hidden.html accessibility/display-contents/search-traversal.html animations/animation-controller-drt-api.html compositing/absolute-inside-out-of-view-fixed.html compositing/accelerated-layers-after-back.html compositing/backing-store-attachment-1.html compositing/backing/backing-store-attachment-animated-transform-inside-fixed.html ... (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46694 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82473 "Found 7 new API test failures: TestWebKitAPI.DocumentEditingContext.SpatialRequestInTextField, TestWebKitAPI.DragAndDropTests.BackgroundImageLinkToInput, TestWebKitAPI.DragAndDropTests.TextAreaToInput, TestWebKitAPI.DragAndDropTests.ImageInLinkToInput, TestWebKitAPI.DragAndDropTests.LinkToInput, TestWebKitAPI.DragAndDropTests.ContentEditableToTextarea, TestWebKitAPI.KeyboardInputTests.SelectionClipRectsWhenPresentingInputView (failure)") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5511 "Found 60 new test failures: accessibility/aria-hidden-subtree.html accessibility/aria-hidden-updates-alldescendants.html accessibility/aria-hidden-with-elements.html accessibility/aria-hidden.html accessibility/aria-roles.html accessibility/display-contents/search-traversal.html accessibility/mac/clipped-text-under-element.html animations/animation-controller-drt-api.html compositing/absolute-inside-out-of-view-fixed.html compositing/accelerated-layers-after-back.html ... (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103943 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23914 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17449 "Found 60 new test failures: accessibility/aria-hidden-subtree.html accessibility/aria-hidden-updates-alldescendants.html accessibility/aria-hidden-with-elements.html accessibility/aria-hidden.html accessibility/display-contents/search-traversal.html animations/animation-controller-drt-api.html compositing/absolute-inside-out-of-view-fixed.html compositing/accelerated-layers-after-back.html compositing/backing-store-attachment-1.html compositing/backing/backing-store-attachment-animated-transform-inside-fixed.html ... (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82846 "Failure limit exceed. At least found 465 new test failures: accessibility/aria-hidden-subtree.html accessibility/aria-hidden-with-elements.html accessibility/aria-hidden.html accessibility/aria-roles.html accessibility/css-content-attribute.html accessibility/gtk/object-attributes.html accessibility/heading-title-includes-links.html accessibility/image-link.html animations/animation-controller-drt-api.html compositing/absolute-inside-out-of-view-fixed.html ... (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/24165 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83683 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82236 "Found 3 new API test failures: /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print-errors, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/close-after-print (failure)") | 
| | [❌ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26898 "Found 9 new test failures: interaction-region/consolidated-nested-regions.html interaction-region/full-page-overlay.html interaction-region/region-area-overflow-does-not-crash.html interaction-region/svg.html interaction-region/text-controls.html interaction-region/text-input-focused-edited-empty.html interaction-region/text-input-focused.html interaction-region/textarea-focused-edited-empty.html interaction-region/textarea-focused.html (failure)") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4448 "Found 60 new test failures: accessibility/aria-hidden-subtree.html accessibility/aria-hidden-updates-alldescendants.html accessibility/aria-hidden-with-elements.html accessibility/aria-hidden.html accessibility/aria-roles.html accessibility/display-contents/search-traversal.html animations/animation-controller-drt-api.html compositing/absolute-inside-out-of-view-fixed.html compositing/accelerated-layers-after-back.html compositing/backing-store-attachment-1.html ... (failure)") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/17408 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23876 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29031 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/23535 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27015 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25276 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/svg/SVGLength.h:
```diff
@@ -48,7 +48,8 @@ class SVGLength : public SVGValueProperty<SVGLengthValue> {
         SVG_LENGTHTYPE_PT           = static_cast<unsigned>(SVGLengthType::Points),
-        SVG_LENGTHTYPE_PC           = static_cast<unsigned>(SVGLengthType::Picas)
+        SVG_LENGTHTYPE_PC           = static_cast<unsigned>(SVGLengthType::Picas),
+        SVG_LENGTHTYPE_REMS           = static_cast<unsigned>(SVGLengthType::Rems)
```
I believe this will be incorrect. `rem` likely needs to return SVG_LENGTHTYPE_UNKNOWN, since it is not defined as part of the IDL (https://www.w3.org/TR/SVG2/types.html#InterfaceSVGLength).

This WPT test seems to imply the same - https://github.com/web-platform-tests/wpt/blob/85b598e652/svg/types/scripted/SVGLength-rem.html.

karlcow replied:
Similar from https://github.com/WebKit/WebKit/pull/22653

https://svgwg.org/svg2-draft/types.html#InterfaceSVGLength
> If other types of lengths are supported and used, the SVGLength uses the SVG_LENGTHTYPE_UNKNOWN unit type. See below for details on how the other properties of an SVGLength operate with these types of lengths.

to add to the commit message.

karlcow replied:
Also the SVG WG has been rechartered recently and a work of maintenance on the specification will happen soon or later (no new features) just streamlining the spec to make sure it describes what is currently implemented where it makes sense.

Ahmad-S792 replied:
I am not sure - this is needed since I did this changed - https://github.com/WebKit/WebKit/commit/e552416b237a8358fc0acbeff8315263139be4a3#diff-c71404f0fc55aa5cae1cf06de7032bc57919bdc529cd4bcac1d430f5e30e4bcaR84

```

     {
         // Per spec: https://svgwg.org/svg2-draft/types.html#__svg__SVGLength__SVG_LENGTHTYPE_UNKNOWN
         if (m_value.lengthType() > SVGLengthType::Picas)
             return 0;
         return static_cast<unsigned>(m_value.lengthType());
     }
```

So technically, it will be treated as `Unknown`.

nullhook replied:
@Ahmad-S792 i was curious why `unitType` was returning 0 even after changing it to a proper value in this file and this check wasn't immediately apparent. do we really need a runtime guard for this? i'd think test failures would be a good indication of not following the spec?

Ahmad-S792 replied:
@nullhook - From IDL, we only have '10; values till SVGLengthType::Picas and when we add anything new, it should return `0` - which is SVGLengthType. For all new length types WPT - if you don't do this. They will fail one of subtests which is if the value is 13, it should return '0'. Without this, it would return '13' and will fail WPT.

karlcow replied:
this is not logical in current STP
https://codepen.io/webcompat/pen/gbOKXxX

`rem` is an invalid value as it is supposed to not be known but still it positionned it correctly.

![Screenshot 2025-03-24 at 10 47 06](https://github.com/user-attachments/assets/b66d104a-7579-4c22-abff-1afcb28681f1)

nullhook replied:
pushed another commit to properly reflect DOMInterface's SVGLength value prop and layout

![image](https://github.com/user-attachments/assets/83568669-eacb-4ec6-a507-90f7fa5d03b3)

nt1m commented on Source/WebCore/svg/SVGLengthContext.cpp:
```diff
@@ -260,6 +264,28 @@ ExceptionOr<float> SVGLengthContext::convertValueFromEMSToUserUnits(float value)
+    if (!style)
+        return Exception { ExceptionCode::NotSupportedError };
+
+    float fontSize = style->computedFontSize();
```
This will be incorrect, we need to get the root element's font size, not the element's

nt1m commented on Source/WebCore/svg/SVGLengthContext.cpp:
```diff
@@ -233,6 +237,19 @@ static inline const RenderStyle* renderStyleForLengthResolving(const SVGElement*
     return nullptr;
 }
 
+static inline const RenderStyle* getRootRenderStyleForLength(const SVGElement* svgElement)
```
```suggestion
static inline const RenderStyle* rootRenderStyleForLength(const SVGElement* svgElement)
```

nit: we don't use the `get` prefix in the webkit codebase

nt1m commented on Source/WebCore/svg/SVGLengthContext.cpp:
```diff
@@ -233,6 +237,19 @@ static inline const RenderStyle* renderStyleForLengthResolving(const SVGElement*
+    if (!svgElement)
+        return nullptr;
+
+    const ContainerNode& node = svgElement->rootNode();
```
This won't do the right thing with shadow roots, you'll want to get the documentElement's render style

nullhook replied:
fixed

nt1m reviewed (commented):
This needs tests

Ahmad-S792 commented on Source/WebCore/svg/SVGLength.h:
```diff
@@ -48,7 +48,8 @@ class SVGLength : public SVGValueProperty<SVGLengthValue> {
         SVG_LENGTHTYPE_PT           = static_cast<unsigned>(SVGLengthType::Points),
-        SVG_LENGTHTYPE_PC           = static_cast<unsigned>(SVGLengthType::Picas)
+        SVG_LENGTHTYPE_PC           = static_cast<unsigned>(SVGLengthType::Picas),
+        SVG_LENGTHTYPE_REMS           = static_cast<unsigned>(SVGLengthType::Unknown)
```
I don't think, it is needed. I didn't added it for `lh` and `ch` length unit and WPT progressed without issue.

nullhook replied:
removed

nullhook committed:
SVGLengthValue.cpp should support rem (probably other) units for width and height attributes
https://bugs.webkit.org/show_bug.cgi?id=258003

Reviewed by NOBODY (OOPS!).

* Source/WebCore/svg/SVGLength.h:
* Source/WebCore/svg/SVGLengthContext.cpp:
(WebCore::SVGLengthContext::convertValueToUserUnits const):
(WebCore::SVGLengthContext::convertValueFromUserUnits const):
(WebCore::SVGLengthContext::convertValueFromUserUnitsToREMS const):
(WebCore::SVGLengthContext::convertValueFromREMSToUserUnits const):
* Source/WebCore/svg/SVGLengthContext.h:
* Source/WebCore/svg/SVGLengthValue.cpp:
(WebCore::lengthTypeToString):
(WebCore::parseLengthType):
(WebCore::primitiveTypeToLengthType):
(WebCore::lengthTypeToPrimitiveType):
* Source/WebCore/svg/SVGLengthValue.h:

karlcow commented on Source/WebCore/svg/SVGLengthValue.h:
```diff
@@ -79,7 +80,7 @@ class SVGLengthValue {
 
     bool isZero() const { return !m_valueInSpecifiedUnits;  }
-    bool isRelative() const { return m_lengthType == SVGLengthType::Percentage || m_lengthType == SVGLengthType::Ems || m_lengthType == SVGLengthType::Exs || m_lengthType == SVGLengthType::Ch; }
+    bool isRelative() const { return m_lengthType == SVGLengthType::Percentage || m_lengthType == SVGLengthType::Ems || m_lengthType == SVGLengthType::Exs || m_lengthType == SVGLengthType::Ch || m_lengthType == SVGLengthType::Rems; }
```
It would be interesting to test how `rem` behaves in other browsers with regards to sizes. Simple SVG test cases should show the differences in the final pixel sizes. 

`isRelative` is used at a number of places
https://searchfox.org/wubkat/search?q=symbol:_ZNK7WebCore14SVGLengthValue10isRelativeEv&redirect=false

nullhook committed:
make sure DOMInterface's SVGLength is same as layout

nullhook committed:
[WIP] SVGRenderStyle must reflect length type

nullhook committed:
[WIP] width and height should respect rems (hack)
